mmf/datasets/processors/bert_processors.py

```diff
@@ -206,3 +212,107 @@ def __call__(self, item):
             processed.segment_ids = processed.segment_ids.view(-1)
             processed.lm_label_ids = processed.lm_label_ids.view(-1)
         return processed.to_dict()
+
+
+@registry.register_processor("masked_multi_sentence_bert_tokenizer")
+class MaskedMultiSentenceBertTokenizer(BertTokenizer):
+    def __init__(self, config, *args, **kwargs):
+        super().__init__(config, *args, **kwargs)
+        self.fusion_strategy = config.get("fusion", "concat")
+        self._probability = getattr(config, "mask_probability", 0.15)
+        self._type = getattr(config, "type", None)
+        self._with_sentence_seperator = getattr(
+            config, "with_sentence_separator", False
+        )
+
+    def __call__(self, item):
+        sentences = item["sentences"]
+        sentences, sentence_tokens = self._tokenize_sentences(sentences)
+        assert len(sentences) == len(sentence_tokens)
+
+        if "two_sentences" in self._type:
+            return self._two_sentences(sentence_tokens)
+        elif "rand_sentences" in self._type:
+            return self._rand_sentences(sentence_tokens)
+        else:  # self._type == "full_sentences":
+            return self._full_sentences(sentence_tokens)
+
+    def _tokenize_sentences(self, texts):
+        sentences, sentence_tokens = [], []
+        for sentence in texts:
+            sentence = sentence.strip()
+            tokens = self._tokenizer.tokenize(sentence)
+            sentences.append(sentence)
+            sentence_tokens.append(tokens)
+        return sentences, sentence_tokens
+
+    def _two_sentences(self, sentence_tokens):
+        """Concatnate only two sentences starting with either 0 or random index."""
+        """With optional separator. Truncate or else pad to max_seq_length."""
+        has_two_or_more_sentences = len(sentence_tokens) > 1
+        if "rand" in self._type:
+            start_idx = (
+                random.randint(0, len(sentence_tokens) - 2)
+                if has_two_or_more_sentences
+                else 0
+            )
+        else:
+            start_idx = 0
+        end_idx = start_idx + 2 if has_two_or_more_sentences else len(sentence_tokens)
+        return self._full_sentences(sentence_tokens[start_idx:end_idx])
+
+    def _rand_sentences(self, sentence_tokens):
+        """Concatnate all consecutive sentences with optional separator."""
+        """Start with a random sentence index aiming to fill max_seq_length."""
+        """Pad till max_seq_length."""
+        sentence_separater = self._sentence_separator("sentence")
+
+        max_start = 0
+        length_to_fullfill = self._max_seq_length - 2
+        for idx, sentence_token in enumerate(sentence_tokens):
+            if sum(len(x) for x in sentence_tokens[idx:]) < length_to_fullfill:
```

Comment:
There is a repetitive calculation here, maybe calculate once, store a counter and subtract?
Also, this code is very hard to read, maybe leave a comment or simplify?